android-components/components/feature/search/src/main/java/mozilla/components/feature/search/SearchUseCases.kt

```diff
@@ -77,16 +77,19 @@ class SearchUseCases(
 
             val id = if (sessionId == null) {
                 // If no `sessionId` was passed in then create a new tab
-                tabsUseCases.addTab(searchUrl)
+                tabsUseCases.addTab(searchUrl, isSearch = true)
             } else {
                 // If we got a `sessionId` then try to find the tab and load the search URL in it
                 val existingTab = store.state.findTabOrCustomTab(sessionId)
                 if (existingTab != null) {
-                    sessionUseCases.loadUrl(searchUrl, sessionId = existingTab.id)
+                    store.dispatch(ContentAction.UpdateIsSearchAction(existingTab.id, true))
+                        .invokeOnCompletion {
```

Comment:
Why do we use `invokeOnCompletion` here, but not in the `else` case where a new tab is added? It's unlikely to be needed here, but if it is, we should also add it to the other case as well?

From the docs of `invokeOnCompletion`: ` There is no guarantee on the execution context in which the handler is invoked.` -> Since these use cases are expected to be invoked on the main thread, I think we will need to wrap the handler in a `launch(Main)`.

Reply:
Ah I see, in the `else` case, the tab use case dispatches an action to the store anyway (to load the URL) and all actions are executed in order so that's why you don't need it there. In `SessionUseCase.loadUrl` on the other hand, you may hit the optimized path that calls `engineSession.loadUrl` directly, which is why you added the `invokeOnCompletion` to make sure the new flag is set before the URL starts loading.

So, yeah, making sure we switch to main for the load inside `invokeOnCompletion` should be good enough then.